src/plugins/analysis/plugin.py

```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
+
+import abc
+import dataclasses
+import time
+import typing
+
+import pydantic
+from packaging.version import Version
+from pydantic import BaseModel, validator
+
+if typing.TYPE_CHECKING:
+    import io
+
+
+class Tag(BaseModel):
+    """A dataclass for tags that is more convenient than dictionaries.
+    The structure of the dict is defined in the docs for :py:attr:`objects.FileObject.analysis_tags`.
+    """
+
+    #: The name of the tag.
+    name: str
+    #: Additional information
+    #: In FACT_core this is shown as tooltip
+    value: str
+    #: The color of the tag
+    #: See :py:class:`helperFunctions.tag.TagColor`.
+    color: str
+    #: Whether or not the tag should be shown in parent files.
+    propagate: bool = False
+
+
+class PluginV0(metaclass=abc.ABCMeta):
+    """An abstract class that all analysis plugins must inherit from.
+
+    Analysis plugins should not depend on FACT_core code where they mustn't.
+    """
+
+    class MetaData(BaseModel):
+        """A class containing all metadata that describes the plugin"""
+
+        #: Name of the plugin
+        name: str
+        #: The plugins description.
+        #: MUST be a single sentence.
+        description: str
+        #: Pydantic model of the object returned by :py:func:`analyse`.
+        # Note that we cannot allow pydantic dataclasses because they lack the `schema` method
+        Schema: typing.Type
+        #: The version of the plugin.
+        #: MUST adhere to PEP 440.
+        #: We suggest using a semver compatible version.
+        version: str
+        #: The version of the backing analysis system.
+        #: E.g. for yara plugins this would be the yara version.
+        system_version: typing.Optional[str] = None
+        #: A list of all plugins that this plugin depends on
+        dependencies: typing.List = pydantic.Field(default_factory=list)
+        #: List of mimetypes that should not be processed
+        mime_blacklist: list = pydantic.Field(default_factory=list)
+        #: List of mimetypes that should be processed
+        mime_whitelist: list = pydantic.Field(default_factory=list)
+        #: The analysis in not expected to take longer than timeout seconds on any given file.
+        timeout: int = 300
+
+        @validator('version')
+        def _version_validator(cls, value):  # noqa: N805
+            _ = Version(value)
+
+            return value
+
+    def __init__(self, metadata: MetaData):
+        self.metadata: PluginV0.MetaData = metadata
```

Comment:
Convert to static variable

Reply:
I'll leave it as it is.
If the metadata is not an argument to the constructor but a static variable it will not complain when the plugin author omits it.